### PR TITLE
Full CRUD of departments, sectors, and users

### PR DESCRIPTION
**Pull request by Snail51**
**BASED ON #15 (NOT main), ADDRESS #15 BEFORE THIS** 

Addresses #3 , #4 
Slight Improvements to #5 
Addresses #13 for these objects (users, sectors, departments)

Full implementation of CRUD for the Users, Sectors, and Departments objects
All routes and functionality have been confirmed to be working on these direct pages.

Note: "Staff Count" field on Departments and Sectors remains unimplemented and is always a "-"

Note: User editing is slightly broken, as removing a user's sector does not clear their department, allowing them to have a department but not a sector. This wont really happen in production but should probably be addressed at some point.

### CODE REVIEW

File: app/Http/Controllers/DepartmentController.php [hunk 1]
```diff
@@ -15,26 +25,60 @@ class DepartmentController extends Controller
...
+        $validated = $request->validate([
+            'name' => ['required','string','max:255'], // required string, max len 255
+            'description' => ['nullable','string','max:255'],  // optional string, max len 255
+            'sector_id' => ['required','integer','exists:sectors,id']     // Ensure 'sector' is a valid integer and exists in the sectors table
+        ]);
+
+        // Programmatically determine the department ID of this new department
+        $lastDepartmentId = Department::latest()->first()->id ?? -1; //if no other department exists, start at -1+1 (0)
+        $newDepartmentId = $lastDepartmentId + 1;
+        $validated['id'] = $newDepartmentId;
```
Comment (drraccoony) [inline app/Http/Controllers/DepartmentController.php]: This isn't needed. primary key ID's with laravel are automatically incrementally generated for us. We dont need to do this. But I see what you were doing.

File: app/Http/Controllers/DepartmentController.php [hunk 2]
```diff
@@ -51,28 +96,86 @@ public function show(string $id)
...
      * Show the form for editing the specified resource.
      */
-    public function edit(string $id)
+    public function edit(Request $request, string $id)
     {
-        $department = Department::findOrFail($id);
-        return view('departments.edit', [
-            'department' => $department
-        ]);
+        if($request->user()->isAdmin())
```
Comment (drraccoony) [inline app/Http/Controllers/DepartmentController.php]: You shouldn't need to do request for the user, remember you should be using `Auth::`. A request can be manipulated to fool the controller into thinking its an admin, whereas the Auth facade is handled on the backend for us with session management.

Refactor this to use `Auth::user`.

File: app/Http/Controllers/DepartmentController.php [hunk 3]
```diff
@@ -51,28 +96,86 @@ public function show(string $id)
...
+            ]);
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     */
+    public function delete(Request $request, string $id): View
     {
-        //
+        if($request->user()->isAdmin())
```
Comment (drraccoony) [inline app/Http/Controllers/DepartmentController.php]: Refactor to use `Auth::user` facade.

File: app/Http/Controllers/SectorController.php [hunk 4]
```diff
@@ -24,30 +24,39 @@ public function index()
     /**
      * Show the form for creating a new resource.
      */
-    public function create()
+    public function create(Request $request)
     {
-        return view('sectors.create');
+        if($request->user()->isAdmin())
```
Comment (drraccoony) [inline app/Http/Controllers/SectorController.php]: Refactor to use `Auth::user`.

File: app/Http/Controllers/SectorController.php [hunk 5]
```diff
@@ -57,35 +66,88 @@ public function store(Request $request)
...
      * Show the form for editing the specified resource.
      */
-    public function edit(string $id)
+    public function edit(Request $request, string $id)
     {
-        $sector = Sector::findOrFail($id);
-        return view('sectors.edit', [
-            'sector' => $sector
-        ]);
+        if($request->user()->isAdmin())
```
Comment (drraccoony) [inline app/Http/Controllers/SectorController.php]: Refactor to use `Auth::user`.

File: app/Http/Controllers/SectorController.php [hunk 6]
```diff
@@ -57,35 +66,88 @@ public function store(Request $request)
...
     }
 
     /**
      * Remove the specified resource from storage.
      */
-    public function destroy(string $id)
+    public function destroy(Request $request, string $id)
     {
-        //
+        if($request->user()->isAdmin())
```
Comment (drraccoony) [inline app/Http/Controllers/SectorController.php]: Refactor to use `Auth::user`

File: app/Http/Controllers/UserController.php [hunk 7]
```diff
@@ -172,8 +172,8 @@ public function delete(Request $request, string $id): View
     {
         if($request->user()->isAdmin())
```
Comment (drraccoony) [inline app/Http/Controllers/UserController.php:186]: This should be `Auth::user`.

File: routes/web.php
Comment (drraccoony) [inline routes/web.php:1]: If `Route::middleware('auth')->group(function () {` is being called multiple times, it wouldn't hurt to nest all routes that depend on this middleware once. Just to clean it up a little.

**Review by drraccoony — CHANGES_REQUESTED** [review]
Good work on the crud calls, just some minor adjustments and its good.

File: app/Http/Controllers/DepartmentController.php (hunk 1 above)
Comment (Snail51) [reply app/Http/Controllers/DepartmentController.php]: resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

File: app/Http/Controllers/DepartmentController.php (hunk 3 above)
Comment (Snail51) [reply app/Http/Controllers/DepartmentController.php]: resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

File: app/Http/Controllers/SectorController.php (hunk 5 above)
Comment (Snail51) [reply app/Http/Controllers/SectorController.php]: resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

File: app/Http/Controllers/SectorController.php (hunk 6 above)
Comment (Snail51) [reply app/Http/Controllers/SectorController.php]: resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

File: app/Http/Controllers/UserController.php (hunk 7 above)
Comment (Snail51) [reply app/Http/Controllers/UserController.php:186]: resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

File: app/Http/Controllers/SectorController.php (hunk 4 above)
Comment (Snail51) [reply app/Http/Controllers/SectorController.php]: resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

File: app/Http/Controllers/DepartmentController.php (hunk 2 above)
Comment (Snail51) [reply app/Http/Controllers/DepartmentController.php]: resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

File: routes/web.php
Comment (Snail51) [reply routes/web.php:1]: resolved in 9c6818f8ae52ff831beb2cc25623d003dcb30c5a

**Pull request merged**